finality-aleph/src/justification.rs

```diff
@@ -1,70 +1,220 @@
 }
 
 impl AlephJustification {
-    pub fn new<Block: BlockT>(auth_crypto_store: &AuthorityKeystore, hash: Block::Hash) -> Self {
-        Self {
-            signature: auth_crypto_store.sign(&hash.encode()[..]),
-        }
+    pub(crate) fn new<Block: BlockT>(signature: SignatureSet<Signature>) -> Self {
+        Self { signature }
     }
 
-    pub(crate) fn _decode_and_verify<Block: BlockT>(
-        justification: &[u8],
+    pub(crate) fn verify<
+        Block: BlockT,
+        MK: MultiKeychain<Signature = Signature, PartialMultisignature = SignatureSet<Signature>>,
+    >(
+        aleph_justification: AlephJustification,
         block_hash: Block::Hash,
-        authorities: &[AuthorityId],
-        number: NumberFor<Block>,
+        multi_keychain: &MK,
     ) -> Result<AlephJustification, Error> {
-        let aleph_justification = AlephJustification::decode(&mut &*justification)
-            .map_err(|_| Error::JustificationDecode)?;
-
-        let encoded_hash = &block_hash.encode()[..];
-        for x in authorities.iter() {
-            if x.verify(&encoded_hash, &aleph_justification.signature) {
-                return Ok(aleph_justification);
-            };
+        let valid =
+            multi_keychain.is_complete(&block_hash.encode()[..], &aleph_justification.signature);
+
+        if !valid {
+            debug!(target: "afa", "Bad justification decoded for block hash #{:?}", block_hash);
+            return Err(Error::BadJustification("Invalid justification".into()));
         }
 
-        log::debug!(target: "afg", "Bad justification decoded for block number #{:?}", number);
-        Err(Error::BadJustification(String::from(
-            "No known AuthorityId was used to sign justification",
-        )))
+        Ok(aleph_justification)
     }
 }
 
-// For now it is glorified proxy channel not doing anything useful
-pub struct JustificationHandler<Block: BlockT> {
-    finalization_proposals_tx: mpsc::UnboundedSender<JustificationNotification<Block>>,
-    justification_rx: mpsc::UnboundedReceiver<JustificationNotification<Block>>,
+pub struct JustificationNotification<Block>
+where
+    Block: BlockT,
+{
+    pub justification: AlephJustification,
+    pub hash: Block::Hash,
+    pub number: NumberFor<Block>,
 }
 
-impl<Block: BlockT> JustificationHandler<Block> {
+type SessionMap<Block> = HashMap<u32, Session<AuthorityId, NumberFor<Block>>>;
+
+pub struct JustificationHandler<B, N, C, BE>
+where
+    B: BlockT,
+    N: network::Network<B> + 'static,
+    C: crate::ClientForAleph<B, BE> + Send + Sync + 'static,
+    BE: Backend<B> + 'static,
+{
+    authority_justification_rx: mpsc::UnboundedReceiver<JustificationNotification<B>>,
+    import_justification_rx: mpsc::UnboundedReceiver<JustificationNotification<B>>,
+    sessions: Arc<Mutex<SessionMap<B>>>,
+    auth_keystore: AuthorityKeystore,
+    session_period: u32,
+    network: N,
+    client: Arc<C>,
+    last_request: Instant,
+    last_finalization: Instant,
```

Comment:
I changed these to `last_request_time` and `last_finalization_time` and I think it's better to leave them as they are. 1) these names are descriptive thus easy to understand when reading, 2) I have changed the condition for requesting justifications, so that the `last_event_time` would not be enough and I expect there will be likely more experimentation with this condition.